mipengine/controller/api/validator.py

```diff
@@ -0,0 +1,312 @@
+import logging
+import numbers
+import traceback
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+
+from mipengine.common.common_data_elements import CommonDataElement
+from mipengine.common.common_data_elements import common_data_elements
+from mipengine.common.node_catalog import node_catalog
+from mipengine.controller.algorithms_specifications import AlgorithmSpecifications
+from mipengine.controller.algorithms_specifications import GenericParameterSpecification
+from mipengine.controller.algorithms_specifications import InputDataSpecification
+from mipengine.controller.algorithms_specifications import InputDataSpecifications
+from mipengine.controller.algorithms_specifications import algorithms_specifications
+from mipengine.controller.api.AlgorithmRequestDTO import AlgorithmInputDataDTO
+from mipengine.controller.api.AlgorithmRequestDTO import AlgorithmRequestDTO
+from mipengine.controller.api.exceptions import BadRequest
+from mipengine.controller.api.exceptions import BadUserInput
+
+
+# TODO This validator will be refactored heavily with https://team-1617704806227.atlassian.net/browse/MIP-68
+# TODO With pydantic remove checks of type "if x is list"
+
+
+def validate_algorithm_request(algorithm_name: str, request_body: str):
+
+    # Validate proper algorithm request body
+    # TODO Should be removed with pydantic
+    try:
+        algorithm_request = AlgorithmRequestDTO.from_json(request_body)
+    except Exception:
+        logging.error(
+            f"Could not parse the algorithm request body. "
+            f"Exception: \n {traceback.format_exc()}"
+        )
+        raise BadRequest(f"The algorithm request body does not have the proper format.")
+
+    algorithm_specs = _get_algorithm_specs(algorithm_name)
+    _validate_algorithm_parameters(algorithm_specs, algorithm_request)
+
+
+def _get_algorithm_specs(algorithm_name):
+    if algorithm_name not in algorithms_specifications.enabled_algorithms.keys():
+        raise BadRequest(f"Algorithm '{algorithm_name}' does not exist.")
+    return algorithms_specifications.enabled_algorithms[algorithm_name]
+
+
+def _validate_algorithm_parameters(
+    algorithm_specs: AlgorithmSpecifications, algorithm_request: AlgorithmRequestDTO
+):
+    _validate_inputdata(algorithm_specs.inputdata, algorithm_request.inputdata)
+
+    _validate_generic_parameters(
+        algorithm_specs.parameters, algorithm_request.parameters
+    )
+
+
+def _validate_inputdata(
+    inputdata_specs: InputDataSpecifications, input_data: AlgorithmInputDataDTO
+):
```

Comment:
Inputdata should be separated between algorithm inputdata and ?application? data. 

This will be done with the algorithms specifications refactoring.

Fixed the input_data and more.